### ISSUE TITLE
Caffe-opencl for Windows

### ISSUE THREAD

**Issue by SHEKOLDA** [BVLC/caffe#4929]
@naibaf7 
Hi, so continuing my quest for unneeded trouble from https://github.com/BVLC/caffe/issues/4872 I tried to build caffe-opencl with Fast R-CNN layers on Windows, which is my sorta 'primary' OS. Since building with CMake wouldn't work as is, I modified it a little, based on the Windows branch. I also decided to build only the core library and python interface (leaving as many of extra tools and tests as possible out) to minimise the number of possible errors to deal with. Still, I'm getting some when building the MSVS solution. They are mostly a result of some syntax errors in automatically generated files.

The errors are:

> error C2059: syntax error : ')' caffe.pb.h    1349    1   caffe
> error C2334: unexpected token(s) preceding ':'; skipping apparent function body caffe.pb.h  1349    1   caffe
> error C2143: syntax error : missing ')' before ';'  caffe.pb.h  1424    1   caffe
> error C2059: syntax error : ')' caffe.pb.h  1424    1   caffe
> error C2238: unexpected token(s) preceding ';'  caffe.pb.h  1424    1   caffe
> error C2027: use of undefined type 'caffe::FillerParameter' generated_message_util.h    100 1   caffe

and then several dozens of similar errors like: 

> error C2071: 'caffe::FillerParameter::FillerParameter_default_instance_' : illegal storage class caffe.pb.h   1425    1   caffe

also these two:

> error C2026: string too big, trailing characters truncated    cl_kernels.cpp  26  1   caffe
> error C2026: string too big, trailing characters truncated  cl_kernels.cpp  33  1   caffe

So I'd love some advice on possible solutions. I'm attaching the txt versions of generated files to reffer to the lines error messegase point to. Thanks in advance!

[caffe.pb.h.txt](https://github.com/BVLC/caffe/files/560792/caffe.pb.h.txt)
[cl_kernels.cpp.txt](https://github.com/BVLC/caffe/files/560795/cl_kernels.cpp.txt)

**Comment by farazfazli** [comment]
Do you have DYLD_LIBRARY_PATH set?

**Comment by naibaf7** [comment]
@SHEKOLDA 
If you could set up a pull-request against OpenCL branch with what you have now, I can see all the code and possibly figure out the issues. 

I appreciate your initiative to try it on your own, I believe this will also save me some time to get this done properly.

**Comment by SHEKOLDA** [comment]
@naibaf7 
Sorry for disappearing like that. I created a PR: https://github.com/BVLC/caffe/pull/4970 and was able to build it yet pycaffe doesn't seem to work.

**Comment by naibaf7** [comment]
@SHEKOLDA great thanks. I might have time to review it this weekend (I need to install Windows and VC2015 first). Stay tuned :)

**Comment by naibaf7** [comment]
@SHEKOLDA @willyd this will be ready in BVLC/opencl today or tomorrow :)

**Comment by willyd** [comment]
@naibaf7 This is great, sorry I did have time to look into this into more details. Are there any changes that should be back ported to the windows branch for easier maintenance?

**Comment by naibaf7** [comment]
@willyd
No, but I might add some improvements to allow regenerating the build files without deleting the whole ```build``` folder as well as multicore compilation if I figure out how to do that.

**Comment by willyd** [comment]
@naibaf7 

> to allow regenerating the build files without deleting the whole build folder

Is that a limitation coming from OpenCL or the way the windows branch is setup? If it's the latter than it might be a good thing to discuss these changes before implementing since it might benefit other people as well. I am asking since some have asked for a more automated way to build the windows branch.  For example, instead of relying on a python script to download the prebuilt dependencies the CMake build could take care of that.

**Comment by naibaf7** [comment]
@willyd 
The changes are not highly complex, I think I'll try what works and we can cherry-pick the good things together after the fact :)
It's currently a build_win.cmd limitation it seems.
The only real change that OpenCL imposes is the necessity to clone viennacl-dev and make the CMake script for finding ViennaCL working. Which is easy.

**Comment by willyd** [comment]
@naibaf7 Ok then. I'll wait for your feedback.

**Comment by willyd** [comment]
@naibaf7 While adding support for python35 I added a CMake script to download the dependencies and remove the need to delete the build folder. Although I did not test it thoroughly the build seems to pass on appveyor. You can cherry pick changes from this branch if it helps you in any way: https://github.com/willyd/caffe/tree/python35

**Comment by naibaf7** [comment]
@willyd 
OK cool.
Status update:
What works:
- Compiling Caffe
- Compiling PyCaffe
- Compiling Gtests
- Running Gtests

What doesn't work:
- On some OpenCL platforms, the binaries just crash on certain layers.

I currently don't know why that happens though, currently under investigation :)

**Comment by naibaf7** [comment]
@willyd @SHEKOLDA 
OpenCL branch supports Windows now.
I've met strange issues in unexpected places... but it should be all good now :)

**Comment by willyd** [comment]
Great news! I may add viennacl to the prebuilt binaries at some point.

**Comment by gfursin** [comment]
When compiling and running OpenCL-based Caffe on Windows (on Intel GPU), I noticed that the startup time can be sometimes 1.5 minutes (and then the classification is quite fast). The problem is in OpenCL kernel caching. Seems like setting the following environment variable can help cache compiled kernels:

> set VIENNACL_CACHE_PATH={SOME PATH WHERE KERNELS WILL BE STORE}
> set CUDA_CACHE_DISABLE=1

See this issue here: https://github.com/dividiti/ck-caffe/issues/44#issuecomment-277205871

I also added ViennaCL-dev package to Collective Knowledge Framework. It should be possible to build and run OpenCL version of Caffe on Windows as following (see https://github.com/dividiti/ck-caffe/wiki/Installation) :
> ck pull repo --url=https://github.com/dividiti/ck-caffe
> ck install package:lib-caffe-bvlc-opencl-viennacl-universal
> ck run program:caffe --env.VIENNACL_CACHE_PATH=C:\Temp\viennacl_cache\ --env.CUDA_CACHE_DISABLE=1

Hope it's of any help ...
